File: 17-letter-combination-of-a-phone-number/ken/17.py
class Solution(object):
	def letterCombinations(self, digits):
		"""
		:type digits: str
		:rtype: List[str]
		"""

		if len(digits) == 0:
			return []

		m = {}
		m['0'] = [' ']
		m['1'] = ['*']
		m['2'] = ['a', 'b', 'c']
		m['3'] = ['d', 'e', 'f']
		m['4'] = ['g', 'h', 'i']
		m['5'] = ['j', 'k', 'l']
		m['6'] = ['m', 'n', 'o']
		m['7'] = ['p', 'q', 'r', 's']
		m['8'] = ['t', 'u', 'v']
		m['9'] = ['w', 'x', 'y', 'z']

		res = ['']
		for digit in digits:
			new_res = []
			for r in res:
				for c in m[digit]:
					new_res.append(r+c)
			res = new_res

		return res

File: 17-letter-combination-of-a-phone-number/leochang/17.cpp
class Solution {
public:
	vector<string> letterCombinations(string digits) {
		vector<string> ret;
		string strMaps[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

		if (0 == digits.size()) {
			return ret;
		}

		ret.push_back("");

		for(int i = 0; i < digits.size(); i++) {
			int num = digits[i] - '0';
			string strChars = strMaps[num];
			vector<string> tmp;
			for (int j = 0; j < ret.size(); j++) {
				for (int k = 0; k < strChars.size(); k++) {
					tmp.push_back(ret[j] + strChars[k]);
				}
			}
			ret = tmp;
		}
		return ret;
	}
};
